app/controllers/flood_risk_engine/enrollments/steps_controller.rb

```diff
@@ -8,30 +8,81 @@
 #
 module FloodRiskEngine
   module Enrollments
-    class StepsController < ApplicationController
-      class StepError < StandardError; end
+    class StepsController < ApplicationController # rubocop:disable Metrics/ClassLength
+
       rescue_from StepError, with: :step_not_found
+
       before_action :check_step_is_valid
       before_action :back_button_cache_buster
+      before_action :clear_error_params, only: [:update, :process_postcode, :process_address]
 
       def show
         form.validate(session[:error_params]) if params[:check_for_error]
+
+        form.prepopulator(params) if form.respond_to?(:prepopulator)
+
         render :show, locals: locals
       end
 
       def update
-        clear_error_params
         success = save_form!
-        if form.redirect?
-          redirect_to(form.redirection_url)
-        elsif success
-          step_forward
-          redirect_to step_url
+
+        # The redirect was happening regardless of outcome, hence moved to within success
+        # i.e if validate or save fails we should NOT redirect
+        if success
+
+          if form.redirect?
```

Comment:
Redirecting only after successful validation breaks the redirect behaviour in previous steps where a redirection was used to exit the process under certain situations. Instead the form should set `form.redirect?` to false if redirection is to be prevented.